File: motion/src/actions/movement.cpp
#include "motion.hpp"

void	Motion::_goToGoal(const turtlesim::Pose &goalPose, const double &distanceTolerance) {

	geometry_msgs::Twist	velMsg;
	double					distance;
	double					desiredAngle;
	ros::Rate				loopRate(100);
	double					kLinear = 0.5;
	double					kAngular = 4.0;

	distance = _getDistance(_turtlesimPose.x, _turtlesimPose.y, goalPose.x, goalPose.y);

	while (distance > distanceTolerance) {
		ROS_INFO("LOOP %.2f, %.2f, %.2f", distance, _getPose().x, _turtlesimPose.y);
		distance = _getDistance(goalPose.y, _turtlesimPose.y, goalPose.x, _turtlesimPose.x);
		velMsg.linear.x = distance * kLinear;
		desiredAngle = atan2(goalPose.y - _turtlesimPose.y, goalPose.x - _turtlesimPose.x);
		velMsg.angular.z = (desiredAngle - _turtlesimPose.theta) * kAngular;

		_velocityPublisher.publish(velMsg);
		ros::spinOnce();
		loopRate.sleep();
	}

	velMsg.linear.x = 0;
	velMsg.angular.z = 0;
	_velocityPublisher.publish(velMsg);
	ROS_INFO("Reached Goal!");
}

void	Motion::_move(double speed, double distance, bool isForward) {

	double					t0;
	double					t1;
	geometry_msgs::Twist	velMsg;
	ros::Rate				loopRate(10);
	double					currentDistance = 0.0;

	if (isForward)
		velMsg.linear.x = abs(speed);
	else
		velMsg.linear.x = -abs(speed);
	velMsg.linear.y = 0;
	velMsg.linear.z = 0;

	t0 = ros::Time::now().toSec();
	while (currentDistance < distance) {
		ROS_INFO("Turtlesim goes forward!");
		_velocityPublisher.publish(velMsg);
		t1 = ros::Time::now().toSec();
		currentDistance = (t1-t0) * speed;
		ros::spinOnce();
		loopRate.sleep();
	}

	ROS_INFO("Arrived!");
	velMsg.linear.x = 0;
	_velocityPublisher.publish(velMsg);
}

  void    Motion::_rotate(double angularSpeed, double relativeAngle, bool clockwise  ) {

	double					t0;
	double					t1;
	geometry_msgs::Twist	velMsg;
	ros::Rate				loopRate(100);
	double					currentAngle = 0.0;

	if (clockwise)
		velMsg.angular.z = -abs(angularSpeed);
	else
		velMsg.angular.z = abs(angularSpeed);

	 t0 = ros::Time::now().toSec();

	 while (currentAngle < relativeAngle) {
		_velocityPublisher.publish(velMsg);
		t1 = ros::Time::now().toSec();
		currentAngle = (t1 - t0) * angularSpeed;
		ros::spinOnce();
		loopRate.sleep();
	}
	ROS_INFO("Reached");
	velMsg.angular.z = 0;
	_velocityPublisher.publish(velMsg);
  }

File: motion/script/actions/movement.py
#!/usr/bin/env python3

import math
import time
import rospy
from turtlesim.msg import Pose
from geometry_msgs.msg import Twist

def goToGoal(velocityPublisher, xGoal, yGoal):

    global x
    global y, yaw

    velocityMessage = Twist()

    while True:
        # kLinear will set the rithm of the speed and distance calculation
        # Can be any value, very high values make the robot unstable
        kLinear = 0.5

        # Calculate the distance to goal
        distance = abs(math.sqrt(((xGoal - x) ** 2) + ((yGoal - y) ** 2)))

        # Robot speed will be proportional to distance for smoothness
        linearSpeed = distance * kLinear

        kAngular = 4.0

        # Mathematically, atan2 gives the angle between two vectors
        desiredAngleGoal = math.atan2(yGoal - y, xGoal - x)
        angularSpeed = (desiredAngleGoal - yaw) * kAngular

        velocityMessage.linear.x = linearSpeed
        velocityMessage.angular.z = angularSpeed

        velocityPublisher.publish(velocityMessage)
        #print('%.3f'%x)
        print('X = %.3f'%x, ', Y = %.3f'%y, ' distance to goal = %.3f'%distance)

        if distance < 0.1:
            break

# angularSpeedDegree = degrees per seccond

def rotate(velocityPublisher, angularSpeedDegree, relativeSpeedDegree, clockwise):

    velocityMessage = Twist()
    angularSpeed = math.radians(abs(angularSpeedDegree))

    if (clockwise):
        velocityMessage.angular.z = -abs(angularSpeed)
    else:
        velocityMessage.angular.z = abs(angularSpeed)

    loopRate = rospy.Rate(100)
    t0 = rospy.Time.now().to_sec()

    while True:
        rospy.loginfo("Turtlesim Rotates")
        velocityPublisher.publish(velocityMessage)

        t1 = rospy.Time.now().to_sec()
        currentAngleDegree = (t1 - t0) * angularSpeedDegree
        loopRate.sleep()

        if currentAngleDegree > relativeSpeedDegree:
            rospy.loginfo("Reached")
            break

    velocityMessage.angular.z = 0
    velocityPublisher.publish(velocityMessage)

def move(velocityPublisher, speed, distance, forward):
    velocityMessage = Twist() # Twist msg to send velocity commands
    # Get current location
    global x, y
    x0 = x # Save the initial location x-coordinate
    y0 = y # Save the initial location y-coordinate

    if forward:
        velocityMessage.linear.x = abs(speed)
    else:
        velocityMessage.linear.x = -abs(speed)

    distanceMoved = 0.0
    loopRate = rospy.Rate(10) # 10Hz

    while True:
        rospy.loginfo("Turtlesim moves forward")
        velocityPublisher.publish(velocityMessage)
        loopRate.sleep()
        distanceMoved = abs(math.sqrt(((x - x0) ** 2) + ((y - y0) ** 2))) # Distance equation -> square root of the square distance traveled point (x, y) minus the square initial point (x0, y0)
        print (distanceMoved)
        if not distanceMoved < distance: # Check if distanceMoved is == to distance
            rospy.loginfo("Arrived!")
            break

    # Stop when distance is met
    velocityMessage.linear.x = 0
    velocityPublisher.publish(velocityMessage)

def poseCallback(poseMessage):
   global x
   global y, yaw
   x = poseMessage.x
   y = poseMessage.y
   yaw = poseMessage.theta

